[inductor] Unify run_and_get_*_code implementations

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112808

The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
[inductor] Unify run_and_get_*_code implementations

The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies. In particular, it pulls in deps that aren't supported on
ROCm.

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Unify run_and_get_*_code implementations"


The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies. In particular, it pulls in deps that aren't supported on
ROCm.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Unify run_and_get_*_code implementations"


The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies. In particular, it pulls in deps that aren't supported on
ROCm.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Unify run_and_get_*_code implementations"


The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies. In particular, it pulls in deps that aren't supported on
ROCm.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/testing/_internal/inductor_utils.py:
```diff
@@ -49,3 +50,39 @@ def _check_has_dynamic_shape(
+        output_code_log.setLevel(logging.DEBUG)
+        result = fn(*args, **kwargs)
+        s = log_capture_string.getvalue()
+        output_code_log.setLevel(prev_level)
```
Nit: could create a context manager for this.

int3 committed:
Update on "[inductor] Unify run_and_get_*_code implementations"


The old implementation of `run_and_get_code` did not work for
AOTInductor. Thus AOTI used `run_and_get_cpp_code` instead. However,
`run_and_get_cpp_code` has nearly the same functionality as
`run_and_get_code`, and with some modification (specifically, splitting
the output code by forward/backward pass), it works for all the use
cases that called the old impl of `run_and_get_code`.

Moreover, I moved the implementation of `run_and_get_cpp_code` out of
test_torchinductor.py because the latter file pulls in many
dependencies.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented on torch/testing/_internal/inductor_utils.py:
```diff
@@ -49,3 +51,46 @@ def _check_has_dynamic_shape(
+    # We use the patch context manager instead of using it as a decorator.
+    # In this way, we can ensure that the attribute is patched and unpatched correctly
+    # even if this run_and_get_code function is called multiple times.
+    with patch.object(config, "debug", True):
```
I know this was already the case with `run_and_get_cpp_code`, but it wasnt the case with `run_and_get_triton_code`. 

Debug does a lot of printing - can we avoid this ? I think it noticeably make tests slower.

int3 replied:
hmm, fair enough, let me try and run some benchmarks.

it looks like `run_and_get_triton_code` is doing a file read, which is not particularly fast either... but that file is probably in disk cache so I guess it might be okay

int3 commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -478,7 +478,7 @@ def func(inp, *, tag, ranks, group_size):
         compiled = torch.compile(func)
         out = compiled(inputs, **self.get_world_trs())
-        code = run_and_get_triton_code(compiled, inputs, **self.get_world_trs())
+        _, code = run_and_get_code(compiled, inputs, **self.get_world_trs())
```
Any suggestions on a better name? Since we're not always returning triton code

int3 commented on test/inductor/test_torchinductor.py:
```diff
@@ -499,30 +477,29 @@ def downcast_fn(x):
-                "tmp" not in stmt,
-                msg=f"Found indirect indexing in statement '{stmt}' from code:\n{code}",
-            )
+    result, code = run_and_get_code(func, *args, **kwargs)
```
Okay, I can switch to `run_and_get_multiple_code` if you would like to keep git blame. I made this change because my initial attempt didn't have that `multiple_code` function, and this test doesn't actually need to split up the source code by file.

eellison replied:
nevermind - i didnt realize there was real changes. thought it was just variable name rename. changes are fine then

int3 commented on test/inductor/test_torchinductor.py:
```diff
@@ -7848,7 +7824,7 @@ def fn():
 
                 fn_opt = torch._dynamo.optimize("inductor")(fn)
-                code = run_and_get_triton_code(fn_opt)
+                _, code = run_and_get_code(fn_opt)
```
I guess you don't like the repeated destructuring... what if I returned a NamedTuple so we could do `code = run_and_capture_code(fn_opt).code`?

Name suggestions for a function that just returns the code would be welcome too
